### PR TITLE
[#3023] doc(core): add event listener document

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?

add event listener document

### Why are the changes needed?

Fix: #3023 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
document

### CODE REVIEW

File: docs/event-listener.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+title: "Gravitino event listener"
+slug: /event-listener
+date: 2024-04-18
+keyword: event listener
+license: Copyright 2024 Datastrato Pvt Ltd. This software is licensed under the Apache License version 2.
+---
+
+## Introduction
+
+Gravitino provides event listener mechanism to allow use to capture the event provides by Gravitino server to inject some custom operations.
```
Comment (jerryshao) [inline docs/event-listener.md:11]: I don't think it is necessary to use a new doc for event listener, you can add a chapter in `gravitino-server-config`

Comment (yuqi1129) [inline docs/event-listener.md:11]: use -> users?

Comment (yuqi1129) [reply docs/event-listener.md:11]: What's the meaning of `inject some custom operations`?

File: docs/event-listener.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+## Event 
+
+Gravitino generates an event following the completion of CREATE, DROP, PURGE, ALTER, UPDATE, LOAD, LIST operations. The events can be hooked to the following resources:
```
Comment (yuqi1129) [inline docs/event-listener.md:15]: What's the difference between `ALTER` and `UPDATE`?

File: docs/event-listener.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+However, events are not hooked to the following resources for now:
+
+- partition
+- role
+- user
+- group
+
+## How to use
+
+To leverage the event listener, you must implement the `EventListenerPlugin` interface and place the resulting JAR in the classpath of the Gravitino server. Then, add configurations to gravitino.conf to enable the event listener.
```
Comment (yuqi1129) [inline docs/event-listener.md:33]: `resulting JAR` --> `resulting JAR files`

File: docs/event-listener.md (hunk 1 above)
Comment (FANNG1) [reply docs/event-listener.md:11]: ok

File: docs/event-listener.md (hunk 3 above)
Comment (FANNG1) [reply docs/event-listener.md]: done

File: docs/event-listener.md (hunk 2 above)
Comment (FANNG1) [reply docs/event-listener.md]: done

File: docs/event-listener.md (hunk 1 above)
Comment (FANNG1) [reply docs/event-listener.md]: rewrited

Comment (FANNG1) [reply docs/event-listener.md]: done

Comment (yuqi1129) [reply docs/event-listener.md]: > use -> users?

You haven't resolved it.

Comment (FANNG1) [reply docs/event-listener.md]: `to allow uses to capture the event`, please check it again

File: docs/gravitino-server-config.md [hunk 4]
```diff
@@ -88,6 +88,50 @@ Gravitino server uses tree lock to ensure the consistency of the data. The tree
 
 Refer to [Iceberg REST catalog service](iceberg-rest-service.md) for configuration details.
 
+### Event listener configuration
+
+Gravitino provides event listener mechanism to allow users to capture the event provides by Gravitino server to integrate some custom operations.
```
Comment (yuqi1129) [inline docs/gravitino-server-config.md:93]: provides -> provided

Comment (yuqi1129) [reply docs/gravitino-server-config.md:93]: From your description, I still can't get what is `event`

Comment (FANNG1) [reply docs/gravitino-server-config.md:93]: I add the description of the event after the configuration, is it enough?

Comment (yuqi1129) [reply docs/gravitino-server-config.md:93]: So, should you put it before the event listener?

File: docs/gravitino-server-config.md [hunk 5]
```diff
@@ -88,6 +88,50 @@ Gravitino server uses tree lock to ensure the consistency of the data. The tree
...
+
+To leverage the event listener, you must implement the `EventListenerPlugin` interface and place the resulting JAR file in the classpath of the Gravitino server. Then, add configurations to gravitino.conf to enable the event listener.
```
Comment (jerryshao) [inline docs/gravitino-server-config.md:95]: What is "resulting", jar is just a jar, why adding "resulting"?

File: docs/gravitino-server-config.md [hunk 6]
```diff
@@ -88,6 +88,50 @@ Gravitino server uses tree lock to ensure the consistency of the data. The tree
...
+#### Event
+
+Gravitino generates an event following the completion of CREATE, DROP, PURGE, ALTER, LOAD, LIST operations. The events can be hooked to the following resources:
+
+- table
+- fileset
+- topic
+- schema
+- catalog
+- metalake
```
Comment (jerryshao) [inline docs/gravitino-server-config.md:112]: You'd better make a table to list all the events for each category.

File: docs/gravitino-server-config.md [hunk 7]
```diff
@@ -88,6 +88,50 @@ Gravitino server uses tree lock to ensure the consistency of the data. The tree
...
+
+However, events are not hooked to the following resources for now:
+
+- partition
+- role
+- user
+- group
```
Comment (jerryshao) [inline docs/gravitino-server-config.md:119]: No need to mention abou this.

File: docs/gravitino-server-config.md [hunk 8]
```diff
@@ -88,6 +88,50 @@ Gravitino server uses tree lock to ensure the consistency of the data. The tree
...
+
+#### Event listener plugin
+
+The `EventListenerPlugin` defines an interface for event listeners that manage the lifecycle and state of a plugin. This includes handling its initialization, startup, and shutdown processes, as well as handing events triggered by various operations.
+
+The plugin provides several operational modes for how to process event, supporting both synchronous and asynchronous processing approaches.
+
+- **SYNC**: Events are processed synchronously, immediately after the related operation is completed. While this method ensures timely event handling, it may block the main process if significant time is required to process an event.
```
Comment (jerryshao) [inline docs/gravitino-server-config.md:127]: what is "timely"?

File: docs/gravitino-server-config.md [hunk 9]
```diff
@@ -88,6 +88,50 @@ Gravitino server uses tree lock to ensure the consistency of the data. The tree
...
+
+- **ASYNC_ISOLATED**: Events are processed asynchronously, with each listener having its own unique event-processing queue and dispatcher thread.
+
+- **ASYNC_SHARED**: In this mode, event listeners utilize a shared event-processing queue and dispatcher to process events asynchronously.
```
Comment (jerryshao) [inline docs/gravitino-server-config.md:131]: What is the difference, pros and cons of these two modes, please clarify it.

File: docs/gravitino-server-config.md (hunk 6 above)
Comment (FANNG1) [reply docs/gravitino-server-config.md]: done

File: docs/gravitino-server-config.md (hunk 7 above)
Comment (FANNG1) [reply docs/gravitino-server-config.md]: removed

File: docs/gravitino-server-config.md (hunk 8 above)
Comment (FANNG1) [reply docs/gravitino-server-config.md]: fixed

File: docs/gravitino-server-config.md (hunk 9 above)
Comment (FANNG1) [reply docs/gravitino-server-config.md]: done

**Pull request merged**